src/controllers/zone_metadata.cr

```diff
@@ -0,0 +1,122 @@
+require "promise"
+
+require "./application"
+
+module PlaceOS::Api
+  class ZoneMetadata < Application
+    include Utils::CoreHelper
+    base "/api/engine/v2/zones/:id"
+
+    before_action :check_support, only: [:update, :destroy]
+    before_action :find_zone
+
+    getter zone : Model::Zone?
+
+    alias Metadata = NamedTuple(
+      name: String?,
+      description: String?,
+      details: JSON::Any?,
+      zone_id: String?
+    )
+
+    get "/metadata", :get_metadata do
+      results = build_metadata(current_zone.metadata, params["name"]?)
+      render json: results
+    end
+
+    get "/children/metadata", :get_children_metadata do
+      filter = params["name"]?
+
+      children = current_zone.children.to_a
+      children.unshift(current_zone)
+
+      results = [] of NamedTuple(
+        zone: Model::Zone,
+        metadata: Hash(String?, Metadata)
+      )
+
+      children.each do |zone|
+        results.push({
+          zone: zone,
+          metadata: build_metadata(zone.metadata, filter)
+        })
+      end
+
+      render json: results
+    end
+
+    post "/metadata", :update do
+      metadata = Metadata.from_json(request.body.as(IO))
+
+      # Metadata shouldn't be nil, use delete instead
+      details = metadata[:details]
+      head :bad_request unless details
+
+      # We need a name to lookup the metadata
+      name = metadata[:name]
+      head :bad_request unless name && !name.empty?
+
+      # TODO:: use database filters
```

Comment:
@caspiano what's the best way to use database filters here?
I'll update them all to match

Reply:
Not optimal but there is a [utility in rethinkdb-orm](https://github.com/spider-gazelle/rethinkdb-orm/blob/master/src/rethinkdb-orm/utils/association_collection.cr) that helps with this.
I can definitely optimise these queries.

```suggestion
    current_zone.metadata.where(name: name).first?
```